diffusion/models/t2i_transformer.py

```diff
@@ -0,0 +1,459 @@
+        # Include flops from adaln
+        param_flops += 2 * self.adaLN_params * batch_size
+        # Calculate flops for attention layers
+        attention_flops = 4 * self.model.num_layers * seq_len**2 * self.model.num_features * batch_size
+        return 3 * param_flops + 3 * attention_flops
+
+    def encode_image(self, image: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Encode an image tensor with the autoencoder and patchify the latents."""
+        with torch.cuda.amp.autocast(enabled=False):
+            latents = self.autoencoder.encode(image.half())['latent_dist'].sample().data
+        # Scale and patchify the latents
+        latents = (latents - self.latent_mean) / self.latent_std
+        latent_patches, latent_coords = patchify(latents, self.patch_size)
+        return latent_patches, latent_coords
+
+    @torch.no_grad()
+    def decode_image(self, latent_patches: torch.Tensor, latent_coords: torch.Tensor) -> torch.Tensor:
+        """Decode image latent patches and unpatchify the image."""
+        # Unpatchify the latents
+        latents = [
+            unpatchify(latent_patches[i], latent_coords[i], self.patch_size) for i in range(latent_patches.shape[0])
+        ]
+        latents = torch.stack(latents)
+        # Scale the latents back to the original scale
+        latents = latents * self.latent_std + self.latent_mean
+        # Decode the latents
+        with torch.cuda.amp.autocast(enabled=False):
+            image = self.autoencoder.decode(latents.half()).sample
+        image = (image / 2 + 0.5).clamp(0, 1)
+        return image
+
+    def tokenize_prompts(self, prompts: Union[str, List[str]]) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Tokenize the prompts using the model's tokenizer."""
+        tokenized_out = self.tokenizer(prompts,
+                                       padding='max_length',
+                                       max_length=self.tokenizer.model_max_length,
+                                       truncation=True,
+                                       return_tensors='pt')
+        return tokenized_out['input_ids'], tokenized_out['attention_mask']
+
+    def combine_attention_masks(self, attention_masks: torch.Tensor) -> torch.Tensor:
+        """Combine attention masks for the encoder if there are multiple text encoders."""
+        if len(attention_masks.shape) == 2:
+            return attention_masks
+        elif len(attention_masks.shape) == 3:
+            encoder_attention_masks = attention_masks[:, 0]
+            for i in range(1, attention_masks.shape[1]):
+                encoder_attention_masks |= attention_masks[:, i]
+            return encoder_attention_masks
+        else:
+            raise ValueError(f'attention_mask should have either 2 or 3 dimensions: {attention_masks.shape}')
+
+    def make_text_embeddings_coords(self, text_embeddings: torch.Tensor) -> torch.Tensor:
```

Comment:
Could this also happen in the `DiffusionTransformer` class?

Reply:
Same as above.